File: api/src/adapters/observatory.ts
/**
 * LLM-Observatory Adapter
 *
 * Thin adapter for consuming telemetry, usage traces, and time-series
 * performance metrics from LLM-Observatory.
 */

import { logger } from '../logger';
import {
  AdapterHealth,
  EcosystemAdapter,
  TimeRange,
  PaginationOptions,
  healthyAdapter,
  unhealthyAdapter,
} from './types';

export interface ObservatoryConfig {
  endpoint: string;
  apiKey?: string;
  timeoutMs?: number;
  batchSize?: number;
}

/**
 * Telemetry data point
 */
export interface TelemetryPoint {
  timestamp: Date;
  metricName: string;
  value: number;
  unit: string;
  tags: Record<string, string>;
  modelId?: string;
  provider?: string;
}

/**
 * Usage trace
 */
export interface UsageTrace {
  traceId: string;
  spanId: string;
  parentSpanId?: string;
  operationName: string;
  startTime: Date;
  endTime: Date;
  durationMs: number;
  status: 'ok' | 'error' | 'timeout';
  attributes: Record<string, unknown>;
  tokenUsage?: TokenUsage;
}

export interface TokenUsage {
  promptTokens: number;
  completionTokens: number;
  totalTokens: number;
}

/**
 * Performance metrics
 */
export interface PerformanceMetrics {
  metricId: string;
  measurement: string;
  timeRange: TimeRange;
  dataPoints: Array<{ timestamp: Date; value: number }>;
  aggregations: MetricAggregations;
}

export interface MetricAggregations {
  min: number;
  max: number;
  avg: number;
  p50: number;
  p95: number;
  p99: number;
  count: number;
}

/**
 * Query parameters for telemetry
 */
export interface TelemetryQuery extends PaginationOptions {
  metricNames?: string[];
  startTime?: Date;
  endTime?: Date;
  modelIds?: string[];
  providers?: string[];
}

/**
 * Query parameters for traces
 */
export interface TraceQuery extends PaginationOptions {
  traceIds?: string[];
  operationNames?: string[];
  startTime?: Date;
  endTime?: Date;
  minDurationMs?: number;
  status?: 'ok' | 'error' | 'timeout';
}

/**
 * LLM-Observatory adapter for consuming telemetry and metrics
 */
export class ObservatoryAdapter implements EcosystemAdapter {
  private config: ObservatoryConfig;
  private connected: boolean = false;

  constructor(config?: Partial<ObservatoryConfig>) {
    this.config = {
      endpoint: process.env.OBSERVATORY_ENDPOINT || 'http://localhost:8081',
      apiKey: process.env.OBSERVATORY_API_KEY,
      timeoutMs: parseInt(process.env.OBSERVATORY_TIMEOUT_MS || '30000', 10),
      batchSize: parseInt(process.env.OBSERVATORY_BATCH_SIZE || '100', 10),
      ...config,
    };
  }

  async connect(): Promise<void> {
    logger.info({ endpoint: this.config.endpoint }, 'Connecting to LLM-Observatory');
    this.connected = true;
    logger.info('Successfully connected to LLM-Observatory');
  }

  async healthCheck(): Promise<AdapterHealth> {
    const start = Date.now();

    if (!this.connected) {
      return unhealthyAdapter('observatory', 'Not connected');
    }

    // In a real implementation, ping the Observatory health endpoint
    const latencyMs = Date.now() - start;
    return healthyAdapter('observatory', latencyMs);
  }

  async disconnect(): Promise<void> {
    logger.info('Disconnecting from LLM-Observatory');
    this.connected = false;
  }

  /**
   * Fetch telemetry data points
   */
  async fetchTelemetry(query: TelemetryQuery): Promise<TelemetryPoint[]> {
    if (!this.connected) {
      throw new Error('Observatory adapter not connected');
    }

    logger.debug({ query }, 'Fetching telemetry from Observatory');

    // Placeholder implementation
    return [];
  }

  /**
   * Fetch usage traces
   */
  async fetchTraces(query: TraceQuery): Promise<UsageTrace[]> {
    if (!this.connected) {
      throw new Error('Observatory adapter not connected');
    }

    logger.debug({ query }, 'Fetching traces from Observatory');

    // Placeholder implementation
    return [];
  }

  /**
   * Fetch time-series performance metrics
   */
  async fetchPerformanceMetrics(
    measurement: string,
    timeRange: TimeRange
  ): Promise<PerformanceMetrics> {
    if (!this.connected) {
      throw new Error('Observatory adapter not connected');
    }

    logger.debug({ measurement, timeRange }, 'Fetching performance metrics from Observatory');

    // Placeholder implementation
    return {
      metricId: crypto.randomUUID(),
      measurement,
      timeRange,
      dataPoints: [],
      aggregations: {
        min: 0,
        max: 0,
        avg: 0,
        p50: 0,
        p95: 0,
        p99: 0,
        count: 0,
      },
    };
  }

  /**
   * Stream telemetry in real-time
   */
  async streamTelemetry(
    metricNames: string[],
    onData: (point: TelemetryPoint) => void
  ): Promise<() => void> {
    if (!this.connected) {
      throw new Error('Observatory adapter not connected');
    }

    logger.info({ metricNames }, 'Starting telemetry stream');

    // In a real implementation, establish WebSocket/SSE connection
    // Return cleanup function
    return () => {
      logger.info('Stopping telemetry stream');
    };
  }
}

File: src/adapters/observatory.rs
//! LLM-Observatory Adapter
//!
//! Thin adapter for consuming telemetry, usage traces, and time-series
//! performance metrics from LLM-Observatory.
//!
//! This adapter provides read-only access to Observatory data for analytics
//! purposes without modifying any upstream logic.

use super::{AdapterHealth, EcosystemAdapter};
use anyhow::{Context, Result};
use async_trait::async_trait;
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::sync::atomic::{AtomicBool, Ordering};
use std::time::Instant;
use tracing::{debug, info, instrument, warn};

/// Configuration for Observatory adapter
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ObservatoryConfig {
    pub endpoint: String,
    pub api_key: Option<String>,
    pub timeout_secs: u64,
    pub batch_size: usize,
}

impl ObservatoryConfig {
    pub fn from_env() -> Result<Self> {
        Ok(Self {
            endpoint: std::env::var("OBSERVATORY_ENDPOINT")
                .unwrap_or_else(|_| "http://localhost:8081".to_string()),
            api_key: std::env::var("OBSERVATORY_API_KEY").ok(),
            timeout_secs: std::env::var("OBSERVATORY_TIMEOUT_SECS")
                .unwrap_or_else(|_| "30".to_string())
                .parse()
                .unwrap_or(30),
            batch_size: std::env::var("OBSERVATORY_BATCH_SIZE")
                .unwrap_or_else(|_| "100".to_string())
                .parse()
                .unwrap_or(100),
        })
    }
}

/// Telemetry data point from Observatory
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TelemetryPoint {
    pub timestamp: DateTime<Utc>,
    pub metric_name: String,
    pub value: f64,
    pub unit: String,
    pub tags: HashMap<String, String>,
    pub model_id: Option<String>,
    pub provider: Option<String>,
}

/// Usage trace from Observatory
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UsageTrace {
    pub trace_id: String,
    pub span_id: String,
    pub parent_span_id: Option<String>,
    pub operation_name: String,
    pub start_time: DateTime<Utc>,
    pub end_time: DateTime<Utc>,
    pub duration_ms: u64,
    pub status: TraceStatus,
    pub attributes: HashMap<String, serde_json::Value>,
    pub token_usage: Option<TokenUsage>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum TraceStatus {
    Ok,
    Error,
    Timeout,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TokenUsage {
    pub prompt_tokens: u64,
    pub completion_tokens: u64,
    pub total_tokens: u64,
}

/// Time-series performance metrics
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PerformanceMetrics {
    pub metric_id: String,
    pub measurement: String,
    pub time_range: TimeRange,
    pub data_points: Vec<DataPoint>,
    pub aggregations: MetricAggregations,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TimeRange {
    pub start: DateTime<Utc>,
    pub end: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DataPoint {
    pub timestamp: DateTime<Utc>,
    pub value: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MetricAggregations {
    pub min: f64,
    pub max: f64,
    pub avg: f64,
    pub p50: f64,
    pub p95: f64,
    pub p99: f64,
    pub count: u64,
}

/// Query parameters for fetching telemetry
#[derive(Debug, Clone, Default)]
pub struct TelemetryQuery {
    pub metric_names: Option<Vec<String>>,
    pub start_time: Option<DateTime<Utc>>,
    pub end_time: Option<DateTime<Utc>>,
    pub model_ids: Option<Vec<String>>,
    pub providers: Option<Vec<String>>,
    pub limit: Option<usize>,
}

/// Query parameters for fetching traces
#[derive(Debug, Clone, Default)]
pub struct TraceQuery {
    pub trace_ids: Option<Vec<String>>,
    pub operation_names: Option<Vec<String>>,
    pub start_time: Option<DateTime<Utc>>,
    pub end_time: Option<DateTime<Utc>>,
    pub min_duration_ms: Option<u64>,
    pub status: Option<TraceStatus>,
    pub limit: Option<usize>,
}

/// LLM-Observatory adapter for consuming telemetry and metrics
pub struct ObservatoryAdapter {
    config: ObservatoryConfig,
    connected: AtomicBool,
}

impl ObservatoryAdapter {
    pub fn new(config: ObservatoryConfig) -> Self {
        Self {
            config,
            connected: AtomicBool::new(false),
        }
    }

    /// Fetch telemetry data points
    #[instrument(skip(self))]
    pub async fn fetch_telemetry(&self, query: TelemetryQuery) -> Result<Vec<TelemetryPoint>> {
        if !self.connected.load(Ordering::Relaxed) {
            anyhow::bail!("Observatory adapter not connected");
        }

        debug!("Fetching telemetry from Observatory");

        // Construct query parameters
        let mut params = HashMap::new();
        if let Some(names) = &query.metric_names {
            params.insert("metrics", names.join(","));
        }
        if let Some(start) = query.start_time {
            params.insert("start", start.to_rfc3339());
        }
        if let Some(end) = query.end_time {
            params.insert("end", end.to_rfc3339());
        }
        if let Some(limit) = query.limit {
            params.insert("limit", limit.to_string());
        }

        // In a real implementation, this would make HTTP calls to Observatory
        // For now, return empty vec as placeholder
        Ok(Vec::new())
    }

    /// Fetch usage traces
    #[instrument(skip(self))]
    pub async fn fetch_traces(&self, query: TraceQuery) -> Result<Vec<UsageTrace>> {
        if !self.connected.load(Ordering::Relaxed) {
            anyhow::bail!("Observatory adapter not connected");
        }

        debug!("Fetching traces from Observatory");

        // In a real implementation, this would make HTTP calls to Observatory
        Ok(Vec::new())
    }

    /// Fetch time-series performance metrics
    #[instrument(skip(self))]
    pub async fn fetch_performance_metrics(
        &self,
        measurement: &str,
        time_range: TimeRange,
    ) -> Result<PerformanceMetrics> {
        if !self.connected.load(Ordering::Relaxed) {
            anyhow::bail!("Observatory adapter not connected");
        }

        debug!(measurement = %measurement, "Fetching performance metrics from Observatory");

        // Placeholder implementation
        Ok(PerformanceMetrics {
            metric_id: uuid::Uuid::new_v4().to_string(),
            measurement: measurement.to_string(),
            time_range,
            data_points: Vec::new(),
            aggregations: MetricAggregations {
                min: 0.0,
                max: 0.0,
                avg: 0.0,
                p50: 0.0,
                p95: 0.0,
                p99: 0.0,
                count: 0,
            },
        })
    }

    /// Stream telemetry in real-time (returns channel receiver)
    pub async fn stream_telemetry(
        &self,
        metric_names: Vec<String>,
    ) -> Result<tokio::sync::mpsc::Receiver<TelemetryPoint>> {
        if !self.connected.load(Ordering::Relaxed) {
            anyhow::bail!("Observatory adapter not connected");
        }

        let (tx, rx) = tokio::sync::mpsc::channel(self.config.batch_size);

        // In a real implementation, this would establish a WebSocket or SSE connection
        // and forward telemetry points through the channel
        info!(metrics = ?metric_names, "Started telemetry stream");

        Ok(rx)
    }
}

#[async_trait]
impl EcosystemAdapter for ObservatoryAdapter {
    #[instrument(skip(self))]
    async fn connect(&self) -> Result<()> {
        info!(endpoint = %self.config.endpoint, "Connecting to LLM-Observatory");

        // In a real implementation, validate connection to Observatory
        self.connected.store(true, Ordering::Relaxed);

        info!("Successfully connected to LLM-Observatory");
        Ok(())
    }

    async fn health_check(&self) -> Result<AdapterHealth> {
        let start = Instant::now();

        if !self.connected.load(Ordering::Relaxed) {
            return Ok(AdapterHealth::unhealthy("observatory", "Not connected"));
        }

        // In a real implementation, ping the Observatory health endpoint
        let latency_ms = start.elapsed().as_millis() as u64;

        Ok(AdapterHealth::healthy("observatory", latency_ms))
    }

    #[instrument(skip(self))]
    async fn disconnect(&self) -> Result<()> {
        info!("Disconnecting from LLM-Observatory");
        self.connected.store(false, Ordering::Relaxed);
        Ok(())
    }
}
